tamr_toolbox/enrichment/address_validation.py

```diff
@@ -0,0 +1,147 @@
+"""Tasks related to validation and refresh of address data using Google Maps API"""
+import logging
+import math
+import os
+from datetime import datetime, timedelta
+from typing import Dict, List, Optional, Tuple
+
+from tamr_toolbox.enrichment.address_mapping import AddressValidationMapping, save, update
+from tamr_toolbox.enrichment.api_client.google_address_validate import validate
+from tamr_toolbox.enrichment.enrichment_utils import join_clean_tuple
+
+# Building our documentation requires access to all dependencies, including optional ones
+# This environments variable is set automatically when `invoke docs` is used
+BUILDING_DOCS = os.environ.get("TAMR_TOOLBOX_DOCS") == "1"
+if BUILDING_DOCS:
+    # Import relevant optional dependencies
+    from googlemaps import Client as GoogleMapsClient
+
+LOGGER = logging.getLogger(__name__)
+
+
+def get_addr_to_validate(
+    input_addresses: List[Tuple[Optional[str], ...]],
+    addr_mapping: Dict[str, AddressValidationMapping],
+    expiration_date_buffer: timedelta = timedelta(days=1),
+) -> List[str]:
+    """Find addresses not previously validated or validated too long ago.
+
+    Args:
+        input_addresses: list of addresses to validate
+        addr_mapping: a dict of address validation mapping data
+        expiration_date_buffer: re-validate addresses if they are within this period of expiring
+
+    Returns:
+        List of standardized addresses not present as keys of the mapping dictionary
+
+    Raises:
+        ValueError: if negative `expiration_date_buffer` is supplied
+    """
+
+    if str(expiration_date_buffer)[0] == "-":
+        raise ValueError("Buffer time for expiration date cannot be negative.")
+
+    count_new_addr = 0
+    count_stale_addr = 0
+    addr_to_validate = []
+
+    for addr in input_addresses:
+        joined_addr = join_clean_tuple(addr)
+        if joined_addr not in addr_mapping.keys():
+            addr_to_validate.append(joined_addr)
+            count_new_addr += 1
+        elif addr_mapping[joined_addr].expiration < str(datetime.now() + expiration_date_buffer):
```

Comment:
Yes, this is all to make it possible to comply with the google terms of service, which prohibit caching most results for more than 30 days. Regarding the first comment, validation won't be deleted if it expires currently, but it's a validation that's still in use (e.g. it's an address that is still in the list of addresses of interest), it will be validated again. 

I don't differentiate between the two cases currently -- I'm not sure there's a reason to do so.